File: disk_speed.java
// disk_speed.java — Java версия

import oshi.SystemInfo;
import oshi.hardware.HardwareAbstractionLayer;
import oshi.hardware.DiskDrive;
import oshi.hardware.DiskStore;

import java.util.*;
import java.util.concurrent.*;

public class disk_speed {
    private static final int INTERVAL = 2;
    private static Map<String, DiskStats> stats = new HashMap<>();

    static class DiskStats {
        List<Double> readSpeeds = new ArrayList<>();
        List<Double> writeSpeeds = new ArrayList<>();
        double maxRead = 0;
        double maxWrite = 0;
    }

    public static void main(String[] args) throws InterruptedException {
        System.out.println("\u001B[36m💾 Disk Speed Monitor (Java)\u001B[0m");
        System.out.println("Интервал: " + INTERVAL + " сек");
        System.out.println("Нажмите Ctrl+C для остановки...");

        SystemInfo si = new SystemInfo();
        HardwareAbstractionLayer hal = si.getHardware();
        List<DiskStore> disks = hal.getDiskStores();

        Map<String, Long> prevRead = new HashMap<>();
        Map<String, Long> prevWrite = new HashMap<>();

        for (DiskStore disk : disks) {
            prevRead.put(disk.getName(), disk.getReadBytes());
            prevWrite.put(disk.getName(), disk.getWriteBytes());
        }

        Thread.sleep(INTERVAL * 1000);

        ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
        executor.scheduleAtFixedRate(() -> {
            try {
                HardwareAbstractionLayer hal2 = si.getHardware();
                List<DiskStore> disks2 = hal2.getDiskStores();

                Map<String, Double> speeds = new HashMap<>();
                Map<String, Long> currRead = new HashMap<>();
                Map<String, Long> currWrite = new HashMap<>();

                for (DiskStore disk : disks2) {
                    currRead.put(disk.getName(), disk.getReadBytes());
                    currWrite.put(disk.getName(), disk.getWriteBytes());
                }

                System.out.println("\n" + "─".repeat(60));
                System.out.printf("\u001B[36m%-6s %-15s %-15s %-10s\u001B[0m\n", "Диск", "Чтение (МБ/с)", "Запись (МБ/с)", "Загрузка");
                System.out.println("─".repeat(60));

                for (DiskStore disk : disks2) {
                    String name = disk.getName();
                    if (prevRead.containsKey(name) && currRead.containsKey(name)) {
                        double readMB = (currRead.get(name) - prevRead.get(name)) / (1024.0 * 1024.0);
                        double writeMB = (currWrite.get(name) - prevWrite.get(name)) / (1024.0 * 1024.0);
                        double readSpeed = readMB / INTERVAL;
                        double writeSpeed = writeMB / INTERVAL;
                        double load = Math.min((readSpeed + writeSpeed) / 1000 * 100, 100);

                        String readColor = readSpeed > 300 ? "\u001B[32m" : readSpeed > 100 ? "\u001B[33m" : "\u001B[31m";
                        String writeColor = writeSpeed > 300 ? "\u001B[32m" : writeSpeed > 100 ? "\u001B[33m" : "\u001B[31m";

                        int barLen = 20;
                        int filled = (int) (load / 100 * barLen);
                        StringBuilder bar = new StringBuilder();
                        for (int i = 0; i < filled; i++) bar.append("█");
                        for (int i = filled; i < barLen; i++) bar.append("░");

                        System.out.printf("%-6s %s%8.1f\u001B[0m     %s%8.1f\u001B[0m     %s %5.0f%%\n",
                            name, readColor, readSpeed, writeColor, writeSpeed, bar.toString(), load);

                        if (!stats.containsKey(name)) {
                            stats.put(name, new DiskStats());
                        }
                        DiskStats ds = stats.get(name);
                        ds.readSpeeds.add(readSpeed);
                        ds.writeSpeeds.add(writeSpeed);
                        if (readSpeed > ds.maxRead) ds.maxRead = readSpeed;
                        if (writeSpeed > ds.maxWrite) ds.maxWrite = writeSpeed;
                    }
                }

                prevRead = currRead;
                prevWrite = currWrite;
            } catch (Exception e) {
                System.err.println("\u001B[31m❌ Ошибка: " + e.getMessage() + "\u001B[0m");
            }
        }, 0, INTERVAL, TimeUnit.SECONDS);

        Thread.sleep(30000); // 30 секунд
        executor.shutdown();
        executor.awaitTermination(5, TimeUnit.SECONDS);

        System.out.println("\n\n\u001B[36m📊 Статистика:\u001B[0m");
        for (Map.Entry<String, DiskStats> entry : stats.entrySet()) {
            DiskStats ds = entry.getValue();
            if (!ds.readSpeeds.isEmpty()) {
                double avgRead = ds.readSpeeds.stream().mapToDouble(Double::doubleValue).average().orElse(0);
                double avgWrite = ds.writeSpeeds.stream().mapToDouble(Double::doubleValue).average().orElse(0);
                System.out.printf("  %s:\n", entry.getKey());
                System.out.printf("    Средняя скорость чтения: %.1f МБ/с\n", avgRead);
                System.out.printf("    Средняя скорость записи: %.1f МБ/с\n", avgWrite);
                System.out.printf("    Пиковая скорость чтения: %.1f МБ/с\n", ds.maxRead);
                System.out.printf("    Пиковая скорость записи: %.1f МБ/с\n", ds.maxWrite);
            }
        }
    }
}

File: disk_speed.rs
// disk_speed.rs — Rust версия

use sysinfo::{System, SystemExt, DiskExt};
use std::{thread, time};
use std::collections::HashMap;

struct DiskStats {
    read_speeds: Vec<f64>,
    write_speeds: Vec<f64>,
    max_read: f64,
    max_write: f64,
}

impl DiskStats {
    fn new() -> Self {
        DiskStats {
            read_speeds: Vec::new(),
            write_speeds: Vec::new(),
            max_read: 0.0,
            max_write: 0.0,
        }
    }
}

fn main() {
    let interval = 2;
    let mut stats: HashMap<String, DiskStats> = HashMap::new();
    let mut prev_read: HashMap<String, u64> = HashMap::new();
    let mut prev_write: HashMap<String, u64> = HashMap::new();

    println!("\x1b[36m💾 Disk Speed Monitor (Rust)\x1b[0m");
    println!("Интервал: {} сек", interval);
    println!("Нажмите Ctrl+C для остановки...");

    let mut sys = System::new_all();
    sys.refresh_all();

    for disk in sys.disks() {
        let name = disk.name().to_str().unwrap_or("unknown").to_string();
        prev_read.insert(name.clone(), disk.total_read_bytes());
        prev_write.insert(name, disk.total_written_bytes());
    }

    thread::sleep(time::Duration::from_secs(interval));

    loop {
        sys.refresh_all();

        let mut curr_read: HashMap<String, u64> = HashMap::new();
        let mut curr_write: HashMap<String, u64> = HashMap::new();

        for disk in sys.disks() {
            let name = disk.name().to_str().unwrap_or("unknown").to_string();
            curr_read.insert(name.clone(), disk.total_read_bytes());
            curr_write.insert(name, disk.total_written_bytes());
        }

        let mut speeds: HashMap<String, (f64, f64)> = HashMap::new();

        for (disk, read) in &curr_read {
            if let Some(prev_r) = prev_read.get(disk) {
                let read_mb = (*read - *prev_r) as f64 / (1024.0 * 1024.0);
                let write_mb = (*curr_write.get(disk).unwrap() - *prev_write.get(disk).unwrap()) as f64 / (1024.0 * 1024.0);
                let read_speed = read_mb / interval as f64;
                let write_speed = write_mb / interval as f64;
                speeds.insert(disk.clone(), (read_speed, write_speed));

                if !stats.contains_key(disk) {
                    stats.insert(disk.clone(), DiskStats::new());
                }
                let s = stats.get_mut(disk).unwrap();
                s.read_speeds.push(read_speed);
                s.write_speeds.push(write_speed);
                if read_speed > s.max_read { s.max_read = read_speed; }
                if write_speed > s.max_write { s.max_write = write_speed; }
            }
        }

        // Вывод
        println!("\n{}", "─".repeat(60));
        println!("\x1b[36mДиск   Чтение (МБ/с)   Запись (МБ/с)   Загрузка\x1b[0m");
        println!("{}", "─".repeat(60));

        for (disk, (read, write)) in &speeds {
            let load = ((read + write) / 1000.0 * 100.0).min(100.0);

            let read_color = if read > 300.0 { "\x1b[32m" } else if read > 100.0 { "\x1b[33m" } else { "\x1b[31m" };
            let write_color = if write > 300.0 { "\x1b[32m" } else if write > 100.0 { "\x1b[33m" } else { "\x1b[31m" };

            let bar_len = 20;
            let filled = (load / 100.0 * bar_len as f64) as usize;
            let bar = format!("{}{}", "█".repeat(filled), "░".repeat(bar_len - filled));

            println!("{:<6} {}{:>8.1}\x1b[0m     {}{:>8.1}\x1b[0m     {} {:>5.0}%",
                disk, read_color, read, write_color, write, bar, load);
        }

        prev_read = curr_read;
        prev_write = curr_write;

        thread::sleep(time::Duration::from_secs(interval));
    }
}
